lib/pure/ospaths.nim

```diff
@@ -678,6 +678,275 @@ when defined(windows) or defined(posix) or defined(nintendoswitch):
       if i > 0: result.add " "
       result.add quoteShell(args[i])
 
+proc isSep(c: char): bool {.noSideEffect.} = c in {DirSep, AltSep}
+
+proc cmpCharInPath(a, b: char): bool {.noSideEffect.} =
+  when FileSystemCaseSensitive:
+    let r = a == b:
+  else:
+    let r = toLowerAscii(a) == toLowerAscii(b)
+  return if r: true else: (a.isSep and b.isSep)
+
+proc sameDrive(a, b: string): bool {.noSideEffect.} =
+  when doslikeFileSystem:
+    not (a.len > 1 and a[1] == ':' and isAlphaAscii(a[0]) and b.len > 1 and b[1] == ':' and a[0] != b[0])
+  else:
+    true
+
+proc countDir(path: string; start, last: Natural): int {.noSideEffect.} =
+  if start >= last:
+    return 0
+
+  result = 0
+  if not path[start].isSep:
+    inc(result)
+  for i in (start+1)..<last:
+    if path[i-1].isSep and not path[i].isSep:
+      inc(result)
+
+proc skipDirSep(path: string; start, last: Natural = 0): int {.noSideEffect.} =
+  var p = start
+  while p < last and path[p].isSep:
+    inc(p)
+  return p
+
+proc rSkipDirSep(path: string; start, last: Natural = 0): int {.noSideEffect.} =
+  var p = start
+  while p > last and path[p].isSep:
+    dec(p)
+  return p
+
+proc countParDir(path: string; start, last: Natural): (int, int) {.noSideEffect.} =
+  var p = start
+  var c = 0
+  while p < last:
+    if p <= last - ParDir.len and continuesWith(path, ParDir, p):
+      p += ParDir.len
+      inc(c)
+      p = skipDirSep(path, p, last)
+    else:
+      break
+  return (c, p)
+
+proc getRelativePathFromAbsolute(path, baseDir: string): string {.
+  noSideEffect.} =
+  ## Convert 'path' to a relative path from baseDir.
+  ##
+  ## Both 'path' and 'baseDir' must be absolute paths.
+  ## On DOS like filesystem, when a drive of 'path' is different from 'baseDir',
+  ## this proc just return the 'path' as is because no way to calculate the relative path.
+  ## This proc never read filesystem.
+  ## 'baseDir' is always assumed to be a directory even if that path is actually a file.
+  ##
+
+  assert(isAbsolute(path) and isAbsolute(baseDir))
+
+  if baseDir.len == 0:
+    return path
+
+  if not sameDrive(path, baseDir):
+    return path
+
+  let alast = path.len
+  let blast = rSkipDirSep(baseDir, baseDir.len - 1, 0) + 1
+
+  var pos = 0
+  let m = min(alast, blast)
+  while pos < m:
+    if not cmpCharInPath(path[pos], baseDir[pos]):
+      break
+    inc(pos)
+
+  if (pos == blast and (alast == blast or path[blast].isSep)) or (pos == alast and (blast > alast and baseDir[pos].isSep)):
+    inc(pos)
+  else:
+    while pos != 0 and not path[pos-1].isSep:
+      dec(pos)
+
+  let numUp = countDir(baseDir, pos, blast)
+
+  if numUp == 0 and pos >= alast:
+    return $CurDir
+
+  result = if numUp > 0: ParDir & (DirSep & ParDir).repeat(numUp-1) else: ""
+  if pos < path.len:
+    return result / path.substr(pos)
```

Comment:
btw if you're gonna go that route (normalize), one way to do it would be to move `os.normalizePath` to `ospaths.normalizePath` (as mentioned in TODO here: https://github.com/nim-lang/Nim/pull/6587#discussion_r200808743)